### PR TITLE
enhance: wal interface definition

### PR DESCRIPTION
**Pull request by chyezh**
issue: #33285

### CODE REVIEW

**Comment by chyezh** [comment]
# WAL

`wal` package is the basic defination of wal interface of milvus lognode.

## Project arrangement

- `/`: only define exposed interfaces.
- `/walimpls/`: define the underlying message system interfaces need to be implemented.
- `/registry/`: A static lifetime registry to regsiter new implementation for inverting dependency.
- `/adaptor/`: adaptors to implement `wal` interface from `walimpls` interface
- `/helper/`: A utility used to help developer to implement `walimpls` conveniently.
- `/utility/`: A utility code for common logic or data structure.

## Lifetime Of Interfaces

- `OpenerBuilder` has a static lifetime in a programs:
- `Opener` keep same lifetime with underlying resources (such as mq client).
- `WAL` keep same lifetime with underlying writer of wal, and it's lifetime is always included in related `Opener`.
- `Scanner` keep same lifetime with underlying reader of wal, and it's lifetime is always included in related `WAL`.

## Add New Implemetation Of WAL

developper who want to add a new implementation of `wal` should implements the `walimpls` package interfaces. following interfaces is required:

- `walimpls.OpenerBuilderImpls`
- `walimpls.OpenerImpls`
- `walimpls.ScannerImpls`
- `walimpls.WALImpls`

`OpenerBuilderImpls` create `OpenerImpls`; `OpenerImpls` creates `WALImpls`; `WALImpls` create `ScannerImpls`. 
Then register the implmentation of `walimpls.OpenerBuilderImpls` into `registry` package.

```
var _ OpenerBuilderImpls = b{};
registry.RegisterBuilder(b{})
```

All things have been done.

## Use WAL

```
name := "your builder name"
var yourCh *logpb.PChannelInfo
opener, err := registry.MustGetBuilder(name).Build()
if err != nil {
    panic(err)
}
ctx := context.Background()
logger, err := opener.Open(ctx, wal.OpenOption{
    Channel: yourCh  
})
if err != nil {
    panic(err)
}
```
## Adaptor

package `adaptor` is used to adapt `walimpls` and `wal` together.
common wal function should be implement by it. Such as:

- lifetime management
- interceptor implementation
- scanner wrapped up
- write ahead cache implementation

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

File: internal/streamingnode/server/wal/walimpls/wal.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/milvus-io/milvus/internal/util/streamingutil/message"
+)
+
+type WALImpls interface {
+	// Channel returns the channel assignment info of the wal.
+	// Should be read-only.
+	Channel() *streamingpb.PChannelInfo
+
+	// Append writes a record to the log.
+	Append(ctx context.Context, msg message.MutableMessage) (message.MessageID, error)
```
Comment (congqixia) [inline internal/streamingnode/server/wal/walimpls/wal.go:16]: Looks like `WALImpls` missing `AppendAsync` method def

**Review by congqixia — APPROVED** [review]
overall lgtm, please sync `WAL` & `WALImpl` interface in next PRs
/lgtm

File: internal/streamingnode/server/wal/walimpls/wal.go (hunk 1 above)
Comment (chyezh) [reply internal/streamingnode/server/wal/walimpls/wal.go:16]: `WALImpls` is not required to implement `AppendAsync`.
`WAL` will give a adaptor to auto implement it.

**Pull request merged**